### PR TITLE
Update: service account for capa and aws provider on irsa

### PR DESCRIPTION
**Pull request by zewolfe**
This PR:
Updates the service account annotation for irsa on `aws` and `capa`

Towards - https://github.com/giantswarm/roadmap/issues/1640

---

## Checklist

- [x] Added a CHANGELOG entry

## Testing

The instance of external-dns installed as part of Giant Swarm platform releases watches services in the `kube-system` namespace with annotations `giantswarm.io/external-dns=managed` and `external-dns.alpha.kubernetes.io/hostname` matching the clusters base domain. (You can find this in the deployments args `--domain-filter` value)

You can take this example `Service`, apply it to your cluster. Change the `external-dns.alpha.kubernetes.io/hostname` annotation to match your clusters base domain.

then:

- Check external-dns logs for lines like `Desired change: CREATE test.your.configured.domain.gigantic.io CNAME`
- Try to resolve the domain (`https://www.dnstester.net/`)

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    external-dns.alpha.kubernetes.io/hostname: test.your.configured.domain.gigantic.io
    external-dns.alpha.kubernetes.io/ttl: "60"
    giantswarm.io/external-dns: managed
  name: test-external-dns
  namespace: kube-system
spec:
  type: ExternalName
  externalName: www.giantswarm.io
```

For testing upgrades:

- Create the service and check for creation
- Upgrade
- Delete the service and check for deletion

### Default app on AWS releases

- [ ] Fresh install works
- [ ] Upgrade works

### Default app on Azure releases

- [ ] Fresh install works
- [ ] Upgrade works

### Optional app (KVM)

- [ ] Fresh install works
- [ ] Upgrade works

### CODE REVIEW

File: helm/external-dns-app/templates/_helpers.tpl [hunk 1]
```diff
@@ -208,7 +208,7 @@ Set Giant Swarm serviceAccountAnnotations.
 */}}
 {{- define "giantswarm.serviceAccountAnnotations" -}}
 {{- if and (or (eq .Values.provider "aws") (eq .Values.provider "capa")) (eq .Values.aws.irsa "true") (eq .Values.aws.access "internal") (not (hasKey .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn")) }}
-{{- $_ := set .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn" (tpl "arn:aws:iam::{{ .Values.aws.accountID }}:role/{{ template \"aws.iam.role\" . }}" .) }}
+{{- $_ := set .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn" (tpl "{{ template \"aws.iam.role\" . }}" .) }}
```
Comment (njuettner) [inline helm/external-dns-app/templates/_helpers.tpl:211]: I think we still need a else if clause here checking if values provider is capa and we also need to remove capa from line 210

File: helm/external-dns-app/values.yaml [hunk 2]
```diff
@@ -15,11 +15,6 @@ aws:
   # can update public and private zones.
   access: internal
 
-  # aws.accountID
-  # AWS account ID is used to assume role via IRSA (IAM roles for service accounts).
-  # It is dynamically set and will be overridden.
-  accountID: ""
```
Comment (njuettner) [inline helm/external-dns-app/values.yaml:21]: we still need this for non vintage clusters

File: helm/external-dns-app/templates/_helpers.tpl [hunk 3]
```diff
@@ -207,8 +207,13 @@ Set Giant Swarm podAnnotations.
 Set Giant Swarm serviceAccountAnnotations.
 */}}
 {{- define "giantswarm.serviceAccountAnnotations" -}}
-{{- if and (or (eq .Values.provider "aws") (eq .Values.provider "capa")) (eq .Values.aws.irsa "true") (eq .Values.aws.access "internal") (not (hasKey .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn")) }}
+{{- if and (eq .Values.aws.irsa "true") (eq .Values.aws.access "internal") (not (hasKey .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn")) }}
+{{- if (eq .Values.provider "aws") }}
 {{- $_ := set .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn" (tpl "arn:aws:iam::{{ .Values.aws.accountID }}:role/{{ template \"aws.iam.role\" . }}" .) }}
+{{- end}}
+{{- if and (or (eq .Values.provider "capa")) }}
```
Comment (njuettner) [inline helm/external-dns-app/templates/_helpers.tpl:214]: We still need to verify aws access internal. The reason is customer can also use "external" which uses ENV to set AWS credentials 

```suggestion
{{- else if and (eq .Values.provider "capa") (eq .Values.aws.access "internal") }}
```

File: helm/external-dns-app/values.schema.json [hunk 4]
```diff
@@ -11,9 +11,6 @@
                 "access": {
                     "type": "string"
                 },
-                "accountID": {
```
Comment (Gacko) [inline helm/external-dns-app/values.schema.json:14]: Can you please add that back? Thank you!

File: helm/external-dns-app/templates/_helpers.tpl [hunk 5]
```diff
@@ -207,8 +207,13 @@ Set Giant Swarm podAnnotations.
...
+{{- $_ := set .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn" (tpl "{{ template \"aws.iam.role\" . }}" .) }}
```
Comment (mcharriere) [inline helm/external-dns-app/templates/_helpers.tpl:215]:
```suggestion
{{- $_ := set .Values.serviceAccount.annotations "eks.amazonaws.com/role-arn" (include "aws.iam.role" .) }}                                                                                                                                        
```

nitpicking here but the `tpl` + `template` is not needed anymore.

File: helm/external-dns-app/values.schema.json (hunk 4 above)
Comment (zewolfe) [reply helm/external-dns-app/values.schema.json:14]: Good catch. Thanks!

File: helm/external-dns-app/templates/_helpers.tpl (hunk 5 above)
Comment (zewolfe) [reply helm/external-dns-app/templates/_helpers.tpl]: 👊🏾

**Review by njuettner — APPROVED** [review]
lgtm

**Review by Gacko — APPROVED** [review]

**Review by mcharriere — APPROVED** [review]
looks good!

**Pull request merged**